node/core/approval-voting/src/lib.rs

```diff
@@ -2760,3 +2837,158 @@ fn issue_local_invalid_statement<Sender>(
 		false,
 	));
 }
+
+async fn get_session_info<'a, Sender>(
+	session_info_provider: &'a mut Option<SessionInfoProvider>,
+	sender: &mut Sender,
+	relay_parent: Hash,
+	session_index: SessionIndex,
+) -> Option<&'a SessionInfo>
+where
+	Sender: SubsystemSender<RuntimeApiMessage>,
+{
+	// If `SessionInfoProvider` is not initialized - try to initialize it first
+	if session_info_provider.is_none() {
+		cache_session_info_for_head(sender, relay_parent, session_info_provider).await;
+	}
+
+	if session_info_provider.is_none() {
+		// `cache_session_info_for_head` should initialize `session_info_provider`
+		// no matter what so log an error if it is still `None`
+		gum::error!(
+			target: LOG_TARGET,
+			session = session_index,
+			?relay_parent,
+			"SessionInfoProvider is not initialized after caching sessions"
+		);
+		return None
+	}
+
+	// And then process the request
+	match session_info_provider
+		.as_mut()
+		.expect("Checked that session_info_provider is Some on the line above. qed.")
+		.runtime_info
+		.get_session_info_by_index(sender, relay_parent, session_index)
+		.await
+	{
+		Ok(extended_info) => Some(&extended_info.session_info),
+		Err(_) => {
+			gum::error!(
+				target: LOG_TARGET,
+				session = session_index,
+				?relay_parent,
+				"Can't obtain SessionInfo"
+			);
+			None
+		},
+	}
+}
+
+/// If `head` is in a new session - cache it
+async fn cache_session_info_for_head<Sender>(
+	sender: &mut Sender,
+	head: Hash,
+	session_info: &mut Option<SessionInfoProvider>,
+) where
+	Sender: SubsystemSender<RuntimeApiMessage>,
+{
+	match session_info.take() {
+		None => {
+			let mut runtime_info = RuntimeInfo::new_with_config(RuntimeInfoConfig {
+				keystore: None,
```

Comment:
Have not checked, but is there really no use for [`ValidatorInfo`](https://github.com/paritytech/polkadot/blob/a7b74ccbfd520146765e20cca9641f7e4ae8e213/node/subsystem-util/src/runtime/mod.rs#L86) in this subsystem? Maybe some code can be simplified by using this?

Reply:
I don't see any usages of `ValidatorInfo` but what simplification you have got in mind?